File: Tablut/src/it/unibo/ai/didattica/competition/tablut/aiclient/algorithms/heuristics/KingSurrounded.java
package it.unibo.ai.didattica.competition.tablut.aiclient.algorithms.heuristics;

import it.unibo.ai.didattica.competition.tablut.aiclient.games.TablutGame;
import it.unibo.ai.didattica.competition.tablut.aiclient.games.board.Coord;
import it.unibo.ai.didattica.competition.tablut.aiclient.games.board.Direction;
import it.unibo.ai.didattica.competition.tablut.aiclient.games.board.StateUtilsKt;
import it.unibo.ai.didattica.competition.tablut.domain.State;

import static it.unibo.ai.didattica.competition.tablut.aiclient.games.board.StateUtilsKt.getKingCoord;

/**
 * An heuristic based on the number of black pawns and obstacles surrounding the king.
 */
public class KingSurrounded implements Heuristic {

    private static final int STEP = 1;

    @Override
    public double evaluate(final TablutGame game, final State state, final State.Turn player) {
        if (game.isTerminal(state)) {
            return game.getUtility(state, player);
        }
        Coord kingPosition = getKingCoord(state);
        double result = 0;
        for (Coord neighbor : kingPosition.coordsAround(KingSurrounded.STEP, state)) {
            State.Pawn neighborPawn = StateUtilsKt.pawnAt(state, neighbor);
            if(neighborPawn == State.Pawn.BLACK || neighborPawn == State.Pawn.THRONE || game.getCitadels().contains(neighbor)) {
                result++;
            }
        }
        if(player == State.Turn.WHITE) {
            return (Direction.values().length - result) / Direction.values().length;
        } else {
            return -(result / Direction.values().length);
        }
    }
}

File: Tablut/src/main/kotlin/it/unibo/ai/didattica/competition/tablut/aiclient/algorithms/heuristics/KingSurrounded.kt
package it.unibo.ai.didattica.competition.tablut.aiclient.algorithms.heuristics

import it.unibo.ai.didattica.competition.tablut.aiclient.game.TablutGame
import it.unibo.ai.didattica.competition.tablut.aiclient.game.board.*
import it.unibo.ai.didattica.competition.tablut.domain.State
import it.unibo.ai.didattica.competition.tablut.domain.State.Pawn
import kotlin.math.min

/**
 * An heuristic based on the number of black and white pawns and obstacles surrounding the king on the row and column
 * where is positioned.
 */
class KingSurrounded : Heuristic {
    private lateinit var citadels: Set<Coord>
    private var maxDistance: Int = 0

    override fun evaluate(game: TablutGame, state: State, player: State.Turn): Double = with (state) {
        if (!::citadels.isInitialized) initialize(game)
        val horizontal = weight(Direction.LEFT) + weight(Direction.RIGHT)
        val vertical = weight(Direction.TOP) + weight(Direction.DOWN)
        0.5 * min(horizontal, vertical) / maxDistance
    }

    private fun State.initialize(game: TablutGame) {
        citadels = game.citadels.toSet()
        maxDistance = size - 2
    }

    /**
     * A white checker has weight +1.
     * A black checker has weight -1.
     * The throne has weight +1 iff it's adjacent to the king, otherwise it's like an empty cell.
     * A citadel has weight -1 iff it's adjacent to the king, otherwise it's like an empty cell.
     * A winning cell has weight +1.
     * An empty cell has weight 0.
     */
    private fun State.weight(direction: Direction): Int {
        var weight = maxDistance
        coordsInDirection(kingCoord, direction).forEach { coord ->
            when (pawnAt(coord)) {
                Pawn.WHITE, Pawn.KING -> return weight
                Pawn.BLACK -> return -weight
                Pawn.THRONE -> return if (weight == maxDistance) weight else 0
                Pawn.EMPTY -> if (citadels.contains(coord)) return if (weight == maxDistance) -weight else 0
            }
            weight--
        }
        return 0
    }
}
